onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -633,87 +707,80 @@ def list(self, request, *args, **kwargs):
 
         return custom_response_handler(request, xform, query, export_type)
 
-    def modify(self, request, *args, **kwargs):
+    @staticmethod
+    def __get_payload(request):
+        try:
+            payload = json.loads(request.data.get('payload', '{}'))
+        except ValueError:
+            raise ParseError(_('Invalid payload'))
 
-        xform = self.get_object()
-        http_status = status.HTTP_200_OK
-        response = {}
+        return payload
+
+    @staticmethod
+    def __get_instances_ids(payload, mongo_base_query):
+
+        """
+        Gets instances ids based on the request payload.
+        Useful to narrow down set of instances for bulk actions
+
+        Args:
+            payload (dict)
+            mongo_base_query (dict): Base query which must contains '_userform_id'
+
+        Returns:
+            list|None: `None` means all instances are returned. Empty list means
+                        no results found
+
+        """
 
-        if request.user.has_perm("validate_xform", xform):
+        instances_ids = None
 
-            owner = xform.user
-            userform_id = "{}_{}".format(owner.username, xform.id_string)
-            query = {ParsedInstance.USERFORM_ID: userform_id}  # Query used for MongoDB
-            filter_ = {"xform_id": xform.id}  # Filter for Django ORM
-            payload = {}
+        ###################################################
+        # Submissions can be retrieve in 3 different ways #
+        ###################################################
 
+        # First scenario / Get submissions based on user's query
+        if payload.get('query'):
+            # Validate if query is valid.
             try:
-                payload = json.loads(request.data.get("payload", "{}"))
+                query = payload.get('query')
+                query.update(mongo_base_query)  # Overrides `_userform_id` if exists
             except ValueError:
```

Comment:
What would raise a `ValueError` here? If someone passes in something other than a JSON object for `query` (which `__get_payload()` converts to a Python `dict`), then `AttributeError` will be raised

Reply:
Forgot to change the Exception. It was because of the former code `json.loads()`.